app/controllers/api/base_controller/renderer.rb

```diff
@@ -470,6 +470,17 @@ def action_validated?(resource, action_spec)
         end
         true
       end
+
+      def render_options(resource, data = {})
+        collection = collection_class(resource)
+        options = {
+          :attributes         => collection.attribute_names,
+          :virtual_attributes => collection.virtual_attribute_names,
+          :relationships      => collection.reflections.keys,
+          :data               => data
+        }
+        render :json => options
+      end
```

Comment:
I'm getting both physical and virtual with the first, and not getting the virtual reflections with the last. Also, we want to make sure we don't advertise encrypted attributes as they are not returned in responses.

```
def render_options(resource, data = {}) 
  collection = collection_class(resource)
  options =
    if collection.blank?
      { :attributes => [], :virtual_attributes => [], :relationships => [] }
    else
      {   
         :attributes => options_attribute_list(collection.attribute_names - collection.virtual_attribute_names),
         :virtual_attributes => options_attribute_list(collection.virtual_attribute_names),
         :relationships => (collection.reflections.keys | collection.virtual_reflections.keys.collect(&:to_s)).sort,
      }   
   end 
   options[:data] = data
   render :json => options
end 
```

```
def options_attribute_list(attrlist)
  attrlist.sort.select { |attr| !encrypted_attribute?(attr) }
end 
```

next method can go in normalizer and you'll need to update normalize_attr_byname to use it so we don't DRY ourselves.

```
def encrypted_attribute?(attr)
   Environment.normalized_attributes[:encrypted].key?(attr.to_s) || attr.to_s.include?("password")
end
```